compiler/infra/ForwardList.hpp

```diff
@@ -0,0 +1,684 @@
+#ifndef FORWARD_LIST_HPP
+#define FORWARD_LIST_HPP
+
+#pragma once
+
+#if 0
```

Comment:
General comment throughout this new file:  I am loathe to introduce large amounts of code that are disabled or commented out without justification.  Are these just debugging artifacts that haven't been cleaned up, future extensions, early experimental implementations, etc.?  I'd rather they not be introduced as I'm doubtful they will ever be cleaned up.